flocker/node/agents/blockdevice.py

```diff
@@ -117,11 +120,73 @@ def __invariant__(self):
                         )
                     )
                 return (False, message)
-        if self.state in (DatasetStates.DELETED,):
+        if self.state in (DatasetStates.DELETED, DatasetStates.NON_EXISTANT):
             return (False, "DesiredDataset can't be in state DELETED.")
         return (True, "")
 
 
+class DesiredDataset(PClass):
+    """
+    Dataset as requested by configuration (and applications).
+    """
+    state = field(
+        invariant=lambda state: (state in DatasetStates.iterconstants(),
+                                 "Not a valid state"),
+        mandatory=True,
+    )
+    dataset_id = field(type=UUID, mandatory=True)
+    maximum_size = field(type=int)
+    metadata = pmap_field(
+        key_type=unicode,
+        value_type=unicode,
+    )
+    mount_point = field(FilePath)
+
+    def __invariant__(self):
+        expected_attributes = [
+            ((DatasetStates.NON_MANIFEST, DatasetStates.MOUNTED),
+             "maximum_size"),
+        ]
+        expected_attributes = [
+            ((DatasetStates.MOUNTED,),
+             "mount_point"),
+        ]
+        for states, attribute in expected_attributes:
+            if (self.state in states) != hasattr(self, attribute):
```

Comment:
`if self.state not in states: continue` would simplify this whole chunk a bit. Ditto for DiscoveredDataset if has same issues.